### PR TITLE
Optional parameters FTW

### PR DESCRIPTION
**Pull request by jskeet**
See #41 for more textual discussion - this is just a PR equivalent of that issue, effectively. (There's room for more PRs for other options such as fluent calls though.)

// cc @mmdriley @chrsmith @remi @SurferJeffAtGoogle

### CODE REVIEW

**Comment by jskeet** [comment]
Oh, and I don't like `default(CancellationToken)` as the default argument value, but we can't use `CancellationToken.None` as that's not a constant value. Have filed a C# 7 feature request :)
https://github.com/dotnet/roslyn/issues/7592

File: CONVENTIONS.md [hunk 1]
```diff
@@ -63,6 +64,14 @@ apply to instance fields, as it's useful to see everything that
 constitutes the state of an object in one place. This is also why
 properties should appear alongside instance fields.
 
+Optional parameters
+---
+
+Optional parameters should be used very sparingly outside the guidelines
+listed here to avoid overload explosion. In particular, any suggested use of
+a default argument value other than the default value for the parameter type is strongly
+discouraged and should be discussed before implementation.
```
Comment (ivannaranjo) [inline CONVENTIONS.md]: Mention the fact that the value used in optional parameters become part of the API and therefore changing it is a breaking change?

**Comment by ivannaranjo** [comment]
Love the deletion of overloads, :shipit:

**Comment by jskeet** [comment]
Added some more comments around breakage - obviously will squish things down before submission.

**Comment by mmdriley** [comment]
Lgtm. Thanks, Jon!

**Comment by ivannaranjo** [comment]
Love it, :shipit:

**Comment by jskeet** [comment]
Have squashed, will merge when green.

**Pull request merged**